assessor/lib/state.rb

```diff
@@ -0,0 +1,55 @@
+
+def state
+  # Memoized
+  unless @state
+    # Look for state only in incremental mode. Do not fail if missing,
+    # just fall back to regular mode, starting with empty state.
+    if @incremental && File.exists?(statepath)
+      @state = YAML.load_file(statepath)
+    else
+      @state = {}
+    end
+    # Schema:
+    # - state[<engine>][<version>]['works']	:: boolean
+    # - state[<engine>][<version>]['app']	:: string
+    #
+    # version is chart version  (`version`)
+    # app     is engine version (`appVersion`)
+    @state.default_proc = proc { |h, k| h[k] = Hash.new }
+  end
+  @state
+end
+
+def state_save(engine, enginev, chartv, success)
+  @state[engine] ||= {}
+  @state[engine][chartv] = {
+    'app'   => enginev,
+    'works' => success,
+  }
+  File.write(statepath, @state.to_yaml)
+end
+
+def archive_save(engine, chartv)
+  dst = File.join(archive_saved, "#{engine}-#{chartv}.tgz")
+  FileUtils.mkdir_p(archive_saved)
+  FileUtils.cp(archive_patched, dst)
+end
+
+def regenerate_working_index
+  # Extract the chart blocks for all working charts from the master,
+  # patch the archive location to refer to the internal dev helm
+  # repository used by the brain tests, and save to a file.
+  working = {}
+  engines.each do |engine|
+    working[engine] = []
+    master_index[engine].each do |chart|
+      chartv = chart['version']
+      next unless @state && @state[engine] && @state[engine][chartv] && @state[engine][chartv]['works']
```

Comment:
```suggestion
      next unless state[engine][chartv] && state[engine][chartv]['works']
```

If you change the [`default_proc`](https://github.com/SUSE/scf-minibroker/pull/1/files#diff-3ca676a5bbdc5ecf2b3e10a42556ef33R18):
```ruby
@state.default_proc = proc { |h, k| h[k] = Hash.new { |hh, kk| hh[kk] = Hash.new } }
```
Then this simplifies to:
```suggestion
      next unless state[engine][chartv]['works']
```

Reply:
Fixed with commit dc68b80856848585810d27f16be4ba547b0cd313.
Not fully simplified, because `['works']` is a third level.
Still have to check `state[engine][chartv]` before accessing that last level.

Reply:
Yep, we'd need to change the `default_proc` for that; it's not important, though, so I'm keeping this resolved.